### DISCUSSION TITLE
Get Inserted/Deleted statements on an Update Query

### DISCUSSION THREAD

**Discussion by daltontc** [eclipse-rdf4j/rdf4j#3163] [Q&A] [answered]
Is there a way that I could retrieve the statements that were modified in an update query? Basically I want to be able to do a "dry-run" of an update query (i.e., start a transaction, run an update, find the changed statements, rollback the transaction). Is something like this even possible?

**Comment by abrokenjester** [comment, marked as answer]
You can do something along those lines with a `SailConnectionListener`. The way to access this is a little bit clunky though. Here's an example:

```java
Repository rep = new SailRepository(new MemoryStore());
try (SailRepositoryConnection conn = (SailRepositoryConnection) rep.getConnection()) {
	NotifyingSailConnection sailConn = (NotifyingSailConnection) conn.getSailConnection();
	sailConn.addConnectionListener(new SailConnectionListener() {

		@Override
		public void statementRemoved(Statement removed) {
			System.out.println("removed: " + removed);
		}

		@Override
		public void statementAdded(Statement added) {
			System.out.println("added: " + added);
		}
	});

	conn.begin();
	conn.add(FOAF.PERSON, RDF.TYPE, RDFS.CLASS);
	String update = "DELETE { ?p a rdfs:Class } INSERT { ?p rdfs:label \"Person\" } WHERE { ?p a rdfs:Class }";
	conn.prepareUpdate(update).execute();
	System.out.println("executed");
	conn.commit();
	System.out.println("transaction committed");
}
```

As you can see we need to cast the `RepositoryConnection` to a specific type to retrieve the underlying `SailConnection`, and then we further need to cast that `SailConnection` to a `NotifyingSailConnection` to be able to register a `SailConnectionListener` on it. This listener will receive pre-commit added and removed events for individual statements. Running the above code would produce the following console output:

```
added: (http://xmlns.com/foaf/0.1/Person, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, http://www.w3.org/2000/01/rdf-schema#Class)
removed: (http://xmlns.com/foaf/0.1/Person, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, http://www.w3.org/2000/01/rdf-schema#Class) [null]
added: (http://xmlns.com/foaf/0.1/Person, http://www.w3.org/2000/01/rdf-schema#label, "Person")
executed
transaction committed
```

**Reply by daltontc** [reply]
Thanks for the response!!! This was super helpful. Is there an easy way using the API to convert an update query into construct queries (one for the inserts and one for the deletions) to get the similar results as above?

**Reply by abrokenjester** [reply]
I don't think there's anything specific to that, but wouldn't a simple regex be enough for that? Just replace `DELETE` (or `INSERT`) with `CONSTRUCT`.

**Reply by daltontc** [reply]
I was originally looking at that but there are quite a lot of edge cases to consider if doing just string manipulation with a generic update query (i.e., separating out if there is both INSERT and DELETE, INSERT DATA, extra brackets in comments, etc.). So thought I'd ask if there was an easy way to do it with ParsedUpdate/ParsedQuery or the SparqlBuilder or another class.

**Reply by abrokenjester** [reply]
Well, I guess you could potentially parse the query and then manipulate the algebra tree to transform it from an update to a query. A INSERT/DELETE query will be represented by a `org.eclipse.rdf4j.query.algebra.Modify` object, which has a `deleteExpr`, a `insertExpr`, and a `whereExpr`, which you can pick out and reuse in a new algebra tree that represent a CONSTRUCT query. The algebra equivalent of a construct clause is a (Multi)Projection on a tuple expression (the statement patterns), mapping variables to predefined aliases for "subject", "predicate" and "object". Play around with the parser a bit, print out some algebra trees to get a feel for how it's structured, and then look into writing a `QueryModelVisitor` implementation that can manipulate the algebra to make it represent a construct query (or just output a new, transformed algebra tree).

**Comment by hugovanversendaal** [comment]
Would it be possible to do this for an HTTPRepository as well? I'm trying to get notifications on updates that take place on an external RDF4J server. I'm trying to implement this using [NotifyingRepositoryWrapper](https://rdf4j.org/javadoc/4.3.16/org/eclipse/rdf4j/repository/event/base/NotifyingRepositoryWrapper.html) with the [addRepositoryListener](https://rdf4j.org/javadoc/4.3.16/org/eclipse/rdf4j/repository/event/base/NotifyingRepositoryWrapper.html#addRepositoryListener(org.eclipse.rdf4j.repository.event.RepositoryListener)) method, but I haven't managed to get there yet. It seems that statementAdded and statementRemoved is not available for RepositoryListener, so if anyone could tell me if what I'm trying achieve is even possible, I would be very grateful.
